docs/rules/0215/foreign-type-reference.md

```diff
@@ -0,0 +1,65 @@
+---
+rule:
+  aip: 215
+  name: [core, '0215', foreign-type-reference]
+  summary: API-specific protos should be in versioned packages.
+permalink: /215/versioned-packages
+redirect_from:
+  - /0215/versioned-packages
+---
+
+# Versioned packages
+
+This rule enforces that none of the fields in an API reference message types in a different
+proto package namespace other than well-known common packages.
+
+## Details
+
+This rule examines all fields in an API's messages and complains if the package of the
+referenced message type is not the same or from one of the common packages such as
+`google.api`, `google.protobuf`, etc.
+
+## Examples
+
+**Incorrect** code for this rule:
+
+```proto
+// Incorrect.
+package foo.bar;
+import "some/other.proto";
+
+message SomeMessage {
+    some.OtherMessage other_message = 1;
+}
+```
+
+**Correct** code for this rule:
+
+```proto
+// Correct.
+package foo.bar;
+import "other.proto";
+
+message SomeMessage {
+    OtherMessage other_message = 1;
+}
+```
+
+## Known issues
+
+This package disallows subpackaging, and allows any package that ends with `.type` to
```

Comment:
> This package disallows subpackaging,

Is this to say that, related to one of Jon's points on #1018 (reproduced below), that `google.library.v1` cannot depend on `google.library.v1.subpkg.Foo` ?

We have some examples of this today that we'd need to support and I think it is still considered an allowable mechanism because it is ostensibly part of the same version.

Jon's relevant comment:
> This should probably be done at an "API version" boundary - so abc/def/v2/xyz/a.proto should be able to import abc/def/v2/b.proto and vice versa

Reply:
The versioned_packages rule already present in the 215 rules defines this regexp:
```
var version = regexp.MustCompile(`\.v[\d]+(p[\d]+)?(alpha|beta|eap|test)?[\d]*`)
```

I think if we normalized the package(s) to the last "version" segment if present we're closer to the intent.  It would mean packages within a version would be allowed, but subpackages without a version segment would be disallowed:

* `foo.v2.a` references `foo.v2.b` -> allowed
* `foo.a` references `foo.b` -> lint warning generated

Reply:
Yeah that makes sense to me!

Reply:
Updated the PR with the change.  I did end up defining a slightly different regex to capture the prefix in one shot, but it does yield an ambiguous case:  is the versioning segment allowed to be the first segment of a package path?

e.g. while `foo.v1` is an allowed path (and thus something like a reference to `foo.v1.bar` is allowed), is `v2` considered a version path if its the only segment, or is it just a package name?

Reply:
> but it does yield an ambiguous case: is the versioning segment allowed to be the first segment of a package path?

This won't realistically happen for Google, but I think the way you've done it would be OK for others